cqrs-postgres/src/reactor.rs

```diff
@@ -0,0 +1,425 @@
+        let mut params: Vec<Box<dyn ToSql>> = Vec::default();
+        let query_with_args = self.generate_query_with_args(R::predicate(), &mut params, 100);
+
+        while self.run.load(Ordering::Relaxed) {
+            let raw_events = {
+                match self.read_all_events(&conn, &query_with_args, since, params.as_slice()) {
+                    Ok(events) => events,
+                    Err(error) => {
+                        panic!(error);
+                    }
+                }
+            };
+
+            for event in raw_events {
+                let event_id = event.event_id;
+                R::react(event).unwrap();
+                since = self
+                    .save_since(&conn, R::reaction_name(), event_id)
+                    .unwrap();
+            }
+
+            ::std::thread::sleep(R::interval());
+        }
+    }
+
+    fn load_since(&self, conn: &Connection, reaction_name: &str) -> Result<Since, postgres::Error> {
+        let trans =
+            conn.transaction_with(postgres::transaction::Config::default().read_only(true))?;
+
+        let rows: Rows = {
+            let stmt = trans.prepare_cached(
+                "SELECT event_id \
+                 FROM reactions \
+                 WHERE reaction_name = $1 \
+                 LIMIT 1",
+            )?;
+            stmt.query(&[&reaction_name])?
+        };
+
+        trans.commit()?;
+
+        for row in rows.iter() {
+            let event_id: Sequence = row.get(0);
+            return Ok(Since::Event(event_id.0));
+        }
+
+        Ok(Since::BeginningOfStream)
+    }
+
+    fn save_since(
+        &self,
+        conn: &Connection,
+        reaction_name: &str,
+        event_id: EventNumber,
+    ) -> Result<Since, postgres::Error> {
+        let trans =
+            conn.transaction_with(postgres::transaction::Config::default().read_only(false))?;
+
+        let stmt = trans.prepare_cached(
+            "INSERT INTO reactions (reaction_name, event_id) \
+             VALUES ($1, $2) \
+             ON CONFLICT (reaction_name) \
+             DO UPDATE SET event_id = EXCLUDED.event_id",
+        )?;
+
+        let rows_updated = stmt
+            .execute(&[
+                &reaction_name,
+                &event_id
+                    .get()
+                    .to_i64()
+                    .expect("Not expecting event_id > several billions"),
+            ])
+            .unwrap();
+
+        trans.commit()?;
+        Ok(Since::Event(event_id))
+    }
+
+    /// Reads all events from the event stream, starting with events after `since`,
+    fn read_all_events(
+        &self,
+        conn: &Connection,
+        query: &str,
+        since: Since,
+        params: &[Box<dyn ToSql>],
+    ) -> Result<Vec<RawEvent>, postgres::Error> {
+        let last_sequence = match since {
+            Since::BeginningOfStream => 0,
+            Since::Event(x) => x.get(),
+        } as i64;
+
+        let trans =
```

Comment:
Is there a specific reason to have a transaction here as opposed to just making the query by itself (Which if it's a single query is treated as a transaction I believe?)

Reply:
IIUMC ("If I Understood Marcus, Correctly"): When we create the `Transaction` with `read_only(true))?;`, PostgreSQL can optimize the transactions better.

Reply:
This is going to go away for some of these, as the single query can be the complete transaction instead.